generate both binary and text annotations

vzaliva opened the pull request:
Due to a known limitation of Tuareg emacs mode https://github.com/ocaml/tuareg/issues/19 always generate .annot files, even if Merlin is present.

**Kind:** infrastructure.

maximedenes commented:
> Due to a known limitation of Tuareg emacs mode

Can you say more about your use case? Can't you use merlin?

vzaliva commented:
This is not critical and many of the functions of Tuareg seems to be covered by Merlin.
Some seem to be not. For example, Tuareg's "show the kind of call at the point",  "show fully qualified ident at the point".

maximedenes commented on configure.ml:
```diff
@@ -390,7 +390,7 @@ let coq_debug_flag = if !Prefs.debug then "-g" else ""
 let coq_annotate_flag =
   if !Prefs.annotate
-  then if program_in_path "ocamlmerlin" then "-bin-annot" else "-annot"
+  then if program_in_path "ocamlmerlin" then "-bin-annot -annot" else "-annot"
```
I would rather like this option to not depend on the programs that are in path at configure-time. Is there any problem with always using `-bin-annot` and `-annot`?

vzaliva replied:
I am not aware of any problems with using both annotations, except slight disk usage to store both.
Let me know if you want me to update the pull request to unconditionally generate both.

ejgallego replied:
`-annot` adds 160 MiB to a build, not something to be taken lightly IMHO.

SkySkimmer replied:
How about we just have both `-bin-annot` and `-annot` as `configure` flags? The user is already passing a flag, we can ask them what exactly they mean. If they want both they can pass both, etc.

vzaliva replied:
I like idea of 2 flags. The only problem is backward compatibility. It is logical not to generate any annotations in absence of any of 2 flags, but it such case it merlin will stop working for all developers until they re-configure with `-bin-annot` flag.

So my suggestion is:
1. Always generate `-bin-annot` regardless of merlin presence. There will be no way to disable this.
2. Optionally `-annot` could be specified to generate text annotations

Zimmi48 replied:
I don't think that backward compatibility is really an issue for developer-centric features.
I'm OK with both solutions: two configure flags, or one that generates both.

vzaliva replied:
I am new to the project and can't decide this. I can tweak the patch either way. I just want to be able to generate text annotations one way or another. If anybody else has an opinion one way or another, please comment.

Zimmi48 replied:
OK, well. If no one else has an opinion to share, I suggest going with @SkySkimmer's proposal. Reflecting the two possible options to `ocamlc` by having the exact same ones in `configure`. The `-annotate` option can be kept just to emit a warning suggesting to use either of these two options instead.

maximedenes replied:
I agree the simplest solution is to reflect the two flags at the `configure` level, and maybe keep the existing one with a deprecation warning.

vzaliva replied:
I've updated the pull requires to have 2 separate flags.

vzaliva commented:
Hmm. This is more that I thought. Perhaps we can add `-anot' flag to configure to control this option.

The question I have is why we currently generate text annotations at all if merlin is not present?

Zimmi48 commented:
> The question I have is why we currently generate text annotations at all if merlin is not present?

Isn't this precisely for tuareg?

Zimmi48 commented on configure.ml:
```diff
@@ -276,6 +276,7 @@ module Prefs = struct
   let flambda_flags = ref []
   let debug = ref true
   let profile = ref false
+  let bin_annotate = ref false
```
This is nitpicking but I would have named the two options `-bin-annot` and `-annot` to reflect exactly the underlying compiler options. If you had used these two names, you could have made `-annotate` emit a warning saying: no op, use `-bin-annot` or `-annot` instead.

vzaliva replied:
I like @Zimmi48 suggestion and implemented it.

Zimmi48 commented on configure.ml:
```diff
@@ -343,7 +344,9 @@ let args_options = Arg.align [
-    " Dumps ml annotation files while compiling Coq";
+    " Dumps ml text annotation files while compiling Coq";
+  "-bin-annotate", Arg.Set Prefs.bin_annotate,
+    " Dumps ml binary annotation files while compiling Coq";
```
Please include in the description some hints on which option is useful for what (Tuareg users vs Merlin users).

Zimmi48 commented on configure.ml:
```diff
@@ -342,8 +343,12 @@ let args_options = Arg.align [
     " Add profiling information in the Coq executables";
-  "-annotate", Arg.Set Prefs.annotate,
-    " Dumps ml annotation files while compiling Coq";
+  "-annotate", Arg.Unit (fun () -> printf "*Warning* -annotate is deprecated Please use -annot or -bin-annot instead\n"),
```
I would add a dot at the end of both sentences, like it is done in other warnings.

Zimmi48 commented:
You forgot to do `git rebase -i HEAD~2 --autosquash` before pushing.

vzaliva commented:
All done now. Sorry, got little confused, first time trying to do rebasing from Magit.

ejgallego commented on configure.ml:
```diff
@@ -388,10 +393,10 @@ let reset_caml_find c o = c.find <- o
+let coq_annotate_flags =
+  String.trim
+    ((if !Prefs.bin_annotate then "-bin-annot" else "") ^
+       " " ^ (if !Prefs.annotate then "-annot" else ""))
```
The right way to do this in OCaml is to use `sprintf`, not `trim`, but anyways this fake flag should be removed and the proper flag added below instead.

ejgallego commented on configure.ml:
```diff
@@ -561,7 +566,7 @@ let coq_warn_error =
 let caml_flags =
-  Printf.sprintf "-thread -rectypes %s %s %s" coq_warnings coq_annotate_flag coq_safe_string
+  Printf.sprintf "-thread -rectypes %s %s %s" coq_warnings coq_annotate_flags coq_safe_string
 
```
Please split this flag in two.

ejgallego commented:
Looks good, thanks for the patch! But please let's keep flags 1-to-1.

vzaliva commented:
I've split the flag into two as requested. Also I adjusted internal variable names to match the flags. The reason I used `trim` was to avoid surplus spaces when flags are not present. This spaces could occur now because of missing this or other flags in `caml_flags`. It is a cosmetic thing easy to fix but I am unsure if you want me to include it in this pull request.

ejgallego commented:
I think it is fine as-is, thanks! Note the squash commit needs squashing again.

vzaliva commented:
squashed.
